File: sdk/android/tracelet-sdk/src/main/kotlin/com/ikolvi/tracelet/sdk/attestation/DeviceAttestor.kt
package com.ikolvi.tracelet.sdk.attestation
import com.ikolvi.tracelet.sdk.util.TraceletLog

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import java.security.MessageDigest
import java.security.SecureRandom
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

/**
 * Device attestation using Google Play Integrity API.
 *
 * Generates signed attestation tokens that prove the device hardware/software
 * is genuine and untampered. Tokens are cached and refreshed periodically.
 *
 * Requires the optional `com.google.android.play:integrity` dependency.
 * If not on the classpath, [requestToken] returns `null` and [isAvailable]
 * returns `false`.
 */
class DeviceAttestor(private val context: Context) {

    companion object {
        private const val TAG = "DeviceAttestor"

        /**
         * Returns `true` if the Play Integrity library is on the classpath.
         */
        @JvmStatic
        fun isAvailable(): Boolean = try {
            Class.forName("com.google.android.play.core.integrity.IntegrityManagerFactory")
            true
        } catch (e: Throwable) {
            false
        }
    }

    // Lazy: avoids NoClassDefFoundError when Play Integrity is absent.
    private val integrityProvider by lazy {
        if (isAvailable()) {
            try {
                Class.forName("com.ikolvi.tracelet.sdk.attestation.PlayIntegrityProvider")
                    .getConstructor(Context::class.java)
                    .newInstance(context) as IntegrityProvider
            } catch (e: Exception) {
                null
            }
        } else null
    }

    private val scheduler = Executors.newSingleThreadScheduledExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())

    @Volatile
    private var cachedToken: Map<String, Any?>? = null

    @Volatile
    private var cachedTimestamp: Long = 0

    private var refreshFuture: ScheduledFuture<*>? = null

    /**
     * Request a fresh attestation token.
     *
     * If a cached token is still valid (within refresh interval), returns
     * the cached version. Otherwise, requests a new token from Play Integrity.
     *
     * @param callback Called with the token map, or null if unavailable
     */
    fun requestToken(callback: (Map<String, Any?>?) -> Unit) {
        if (!isAvailable()) {
            TraceletLog.warning("Play Integrity not available. " +
                "Add implementation(\"com.google.android.play:integrity:1.6.0\") " +
                "to your app's build.gradle.")
            callback(null)
            return
        }

        // Return cached token if still fresh (within 5 minutes)
        val cached = cachedToken
        if (cached != null && System.currentTimeMillis() - cachedTimestamp < 300_000) {
            callback(cached)
            return
        }

        val nonce = generateNonce()
        try {
            integrityProvider?.requestToken(nonce,
                onSuccess = { token ->
                    val result = mapOf<String, Any?>(
                        "token" to token,
                        "timestamp" to System.currentTimeMillis(),
                        "provider" to "play_integrity",
                        "verified" to null, // Server-side verification needed
                    )
                    cachedToken = result
                    cachedTimestamp = System.currentTimeMillis()
                    callback(result)
                },
                onFailure = { e ->
                    TraceletLog.warning("Play Integrity request failed: ${e.message}")
                    callback(null)
                }
            ) ?: run {
                callback(null)
            }
        } catch (e: Exception) {
            TraceletLog.warning("Play Integrity unavailable: ${e.message}")
            callback(null)
        }
    }

    /**
     * Start periodic token refresh.
     *
     * @param intervalSeconds Refresh interval in seconds (minimum 60)
     */
    fun startRefresh(intervalSeconds: Int) {
        if (!isAvailable()) return
        stopRefresh()
        val interval = intervalSeconds.coerceAtLeast(60).toLong()
        refreshFuture = scheduler.scheduleAtFixedRate(
            { requestToken { /* cache update only */ } },
            interval,
            interval,
            TimeUnit.SECONDS
        )
    }

    /** Stop periodic token refresh. */
    fun stopRefresh() {
        refreshFuture?.cancel(false)
        refreshFuture = null
    }

    /** Get the last cached attestation token, or null. */
    fun getCachedToken(): Map<String, Any?>? = cachedToken

    /**
     * Generate a unique nonce for the integrity request.
     * SHA-256 of device-specific data + timestamp + random bytes.
     */
    private fun generateNonce(): String {
        val random = ByteArray(16)
        SecureRandom().nextBytes(random)
        val data = "${android.os.Build.FINGERPRINT}:${System.currentTimeMillis()}:${random.joinToString("") { "%02x".format(it) }}"
        val digest = MessageDigest.getInstance("SHA-256").digest(data.toByteArray())
        return digest.joinToString("") { "%02x".format(it) }
    }
}

interface IntegrityProvider {
    fun requestToken(nonce: String, onSuccess: (String) -> Unit, onFailure: (Exception) -> Unit)
}

File: sdk/ios/Sources/TraceletSDK/attestation/DeviceAttestor.swift
import Foundation
import DeviceCheck
import CryptoKit

/// Device attestation using Apple's App Attest API (iOS 14+).
///
/// Generates hardware-backed attestation tokens that prove the device
/// is genuine and the app binary has not been tampered with.
/// Tokens are cached and periodically refreshed.
public final class DeviceAttestor {

    private var cachedToken: [String: Any]?
    private var cachedTimestamp: Date?
    private var refreshTimer: Timer?
    private let cacheDuration: TimeInterval = 300 // 5 minutes

    /// Request an attestation token.
    ///
    /// Returns a cached token if still fresh, otherwise generates a new one.
    ///
    /// - Parameter completion: Called with the token dictionary, or nil if unavailable.
    func requestToken(completion: @escaping ([String: Any]?) -> Void) {
        // Return cached token if still fresh
        if let cached = cachedToken,
           let ts = cachedTimestamp,
           Date().timeIntervalSince(ts) < cacheDuration {
            completion(cached)
            return
        }

        guard #available(iOS 14.0, *) else {
            completion(nil)
            return
        }

        let service = DCAppAttestService.shared
        guard service.isSupported else {
            TraceletLog.debug("[DeviceAttestor] App Attest not supported on this device")
            completion(nil)
            return
        }

        service.generateKey { [weak self] keyId, error in
            guard let keyId = keyId else {
                TraceletLog.error("[DeviceAttestor] Key generation failed: \(error?.localizedDescription ?? "unknown")")
                completion(nil)
                return
            }

            let challenge = self?.generateChallenge() ?? Data()
            let hash = Data(SHA256.hash(data: challenge))

            service.attestKey(keyId, clientDataHash: hash) { [weak self] attestation, error in
                guard let attestation = attestation else {
                    TraceletLog.error("[DeviceAttestor] Attestation failed: \(error?.localizedDescription ?? "unknown")")
                    completion(nil)
                    return
                }

                let result: [String: Any] = [
                    "token": attestation.base64EncodedString(),
                    "timestamp": Int(Date().timeIntervalSince1970 * 1000),
                    "provider": "app_attest",
                    "verified": NSNull(),
                ]
                self?.cachedToken = result
                self?.cachedTimestamp = Date()
                completion(result)
            }
        }
    }

    /// Start periodic token refresh.
    ///
    /// - Parameter intervalSeconds: Refresh interval in seconds (minimum 60).
    func startRefresh(intervalSeconds: Int) {
        stopRefresh()
        let interval = max(60, intervalSeconds)
        DispatchQueue.main.async { [weak self] in
            self?.refreshTimer = Timer.scheduledTimer(withTimeInterval: TimeInterval(interval), repeats: true) { [weak self] _ in
                self?.requestToken { _ in /* cache update */ }
            }
        }
    }

    /// Stop periodic token refresh.
    func stopRefresh() {
        refreshTimer?.invalidate()
        refreshTimer = nil
    }

    /// Returns the last cached attestation token, or nil.
    func getCachedToken() -> [String: Any]? {
        return cachedToken
    }

    /// Generate a challenge for the attestation request.
    private func generateChallenge() -> Data {
        var bytes = [UInt8](repeating: 0, count: 32)
        _ = SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes)
        return Data(bytes)
    }
}
